network/gossip/libp2p/topology/statefulTopologyManager.go

```diff
@@ -0,0 +1,83 @@
+package topology
+
+import (
+	"fmt"
+
+	"github.com/onflow/flow-go/engine"
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/network/gossip/libp2p/channel"
+)
+
+// StatefulTopologyManager implements Manager interface.
+// The term stateful denotes the fact that it tries to optimize fanout by constructing topology state-by-state.
+// Where the next state of topology utilizes nodes chosen in the current state as much as possible, hence to
+// avoid unnecessary redundancies.
+type StatefulTopologyManager struct {
+	topology Topology                    // used to sample nodes
+	subMngr  channel.SubscriptionManager // used to keep track topics the node subscribed to
+}
+
+// NewStatefulTopologyManager generates and returns an instance of stateful topology manager.
+func NewStatefulTopologyManager(topology Topology, subMngr channel.SubscriptionManager) *StatefulTopologyManager {
+	return &StatefulTopologyManager{
+		topology: topology,
+		subMngr:  subMngr,
+	}
+}
+
+// MakeTopology receives identity list of entire network and constructs identity list of topology
+// of this instance. A node directly communicates with its topology identity list on epidemic dissemination
+// of the messages (i.e., publish and multicast).
+// Independent invocations of MakeTopology on different nodes collaboratively must construct a cohesive
+// connected graph of nodes that enables them talking to each other.
+func (stm *StatefulTopologyManager) MakeTopology(ids flow.IdentityList) (flow.IdentityList, error) {
+	myChannelIDs := stm.subMngr.GetChannelIDs()
+	if len(myChannelIDs) == 0 {
+		// no subscribed channel id, hence skip topology creation
+		// we do not return an error at this state as invocation of MakeTopology may happen before
+		// node subscribing to all its channels.
+		return flow.IdentityList{}, nil
+	}
+
+	// finds all interacting roles with this node
+	myInteractingRoles := flow.RoleList{}
+	for _, myChannel := range myChannelIDs {
+		roles, ok := engine.RolesByChannelID(myChannel)
+		if !ok {
+			return nil, fmt.Errorf("could not extract roles for channel: %s", myChannel)
+		}
+		myInteractingRoles = myInteractingRoles.Union(roles)
+	}
+
+	// builds a connected component per role this node interact with,
+	var myFanout flow.IdentityList
+	for _, role := range myInteractingRoles {
+		if role == flow.RoleCollection {
+			// we do not build connected component for collection nodes based on their role
+			// rather we build it based on their cluster identity in the next step.
+			continue
+		}
+		roleFanout, err := stm.topology.SubsetRole(ids, nil, flow.RoleList{role})
+		if err != nil {
+			return nil, fmt.Errorf("failed to derive list of peer nodes to connect for role %s: %w", role, err)
+		}
+		myFanout = myFanout.Union(roleFanout)
+	}
+
+	// stitches the role-based components that subscribed to the same channel id together.
+	for _, myChannel := range myChannelIDs {
```

Comment:
Why do this if we have already connected the nodes by their roles (except for Collection nodes)?

Reply:
☝️  I meant, if are choosing by roles, then why again choose by channel-id?

Reply:
With the first for loop, the system is partitioned based on roles. However, inter-role connections (i.e., channels) are not necessarily guaranteed. The second loop takes care of that by connecting the _interacting_ roles. This is basically a stepping stone for the next step where we implement a randomized graph sampler (we do not change the topology manager). Then the first loop connects roles _with some probability (p)_, and the second loop connected channels with _some probability_, so the system is connected with high probability, i.e., probability in the form of 1-p^-2.

Reply:
If there is a topic A which is subscribed by lets say the AN and the EN, then if we connect by role, ANs and ENs will be connected directly or indirectly. Then why worry about connecting them again by Topic A?

Reply:
Marking this is as resolved after speaking with @yhassanzadeh13. 
It has been observed that if only topics are used to select nodes, many a times we end up with an unequal distribution of nodes by role for the given topic. e.g. if there are 50 Consensus nodes and 50 collection nodes, and we choose randomly for a topic that involves only these two roles, then we can end up with say 2 consensus nodes and 49 collection node. This happens because the seed is not random and to converge on an truly random distribution where there is 50% chance of choosing consensus node and 50% change of choosing a collection node,  we will have to make the choice exponential number of times.